config/src/lib.rs

```diff
@@ -1,75 +1,113 @@
+//! load the configuration as an instance of the `Config` struct which
+//! is composed of other, more specialized, structs such as
+//! `StorageConfig` and `ConnectionsConfig`. This instance is the one
+//! you use in your Rust code to interact with the loaded
+//! configuration.
+#![cfg_attr(test, allow(dead_code, unused_macros, unused_imports))]
 
 #[macro_use]
 extern crate serde_derive;
 
-#[derive(Debug, Deserialize)]
+use std::default::Default;
+use std::net::SocketAddr;
+
+pub mod loaders;
+
+/// The entire configuration
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
 pub struct Config {
-    pub server: ServerConfig,
-    pub logging: Option<LoggingConfig>,
+    #[serde(default = "Config::default_connections")]
+    pub connections: ConnectionsConfig,
+    #[serde(default = "Config::default_storage")]
+    pub storage: StorageConfig,
 }
 
-#[derive(Debug, Deserialize)]
-pub struct ServerConfig {
-    pub chain_type: Option<String>,
-    pub db_root: Option<String>,
-    pub host: String,
-    pub p2p: Option<P2pConfig>,
-    pub port: u16,
+/// Connections-specific configuration
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct ConnectionsConfig {
+    #[serde(default = "ConnectionsConfig::default_server_addr")]
+    pub server_addr: SocketAddr,
+    #[serde(default = "ConnectionsConfig::default_inbound_limit")]
+    pub inbound_limit: u16,
+    #[serde(default = "ConnectionsConfig::default_outbound_limit")]
+    pub outbound_limit: u16,
+    #[serde(default = "ConnectionsConfig::default_known_peers")]
+    pub known_peers: Vec<SocketAddr>,
 }
 
-#[derive(Clone, Debug, Deserialize)]
-pub struct P2pConfig {
-    pub host: String,
-    pub port: u16,
+/// Storage-specific configuration
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct StorageConfig {
+    #[serde(default = "StorageConfig::default_db_path")]
+    pub db_path: String,
 }
 
-#[derive(Debug, Deserialize)]
-pub struct LoggingConfig {
-    pub file_log_level: Option<String>,
-    pub log_file_append: Option<bool>,
-    pub log_file_path: Option<String>,
-    pub log_to_file: Option<bool>,
-    pub log_to_stdout: Option<bool>,
-    pub stdout_log_level: Option<String>,
+impl Config {
+    fn default_connections() -> ConnectionsConfig {
+        ConnectionsConfig::default()
+    }
+
+    fn default_storage() -> StorageConfig {
+        StorageConfig::default()
+    }
+}
+
+impl Default for Config {
+    fn default() -> Config {
+        Config {
+            connections: Self::default_connections(),
+            storage: Self::default_storage(),
+        }
+    }
 }
 
-pub fn read_config() -> Option<Config> {
-    let name: String = String::from(CONFIG_FILE);
-    let mut input = String::new();
+impl ConnectionsConfig {
+    fn default_server_addr() -> SocketAddr {
+        "127.0.0.1:1234".parse().unwrap()
+    }
+
+    fn default_inbound_limit() -> u16 {
+        256
```

Comment:
I think that we have usually talked about a smaller default limit (around 128). Please talk to the team to check the concrete number.

Reply:
AFAIK, for now we are being inspired by Bitcoin parameters. They have 117 inbound peers, so we could set the limit to something similar (e.g. 128 if we want to have a power of two number :smile: ).

Reply:
The outbound limit is correct! :+1: